File: dev/dumper_binja/type_parser.py
"""
Type decomposition — converts raw C++ type strings from RTTI into structured
field descriptors that downstream code generators can consume without parsing
C++ template syntax.

Example:
    decompose_type("class std::optional<class std::vector<struct Api::OneMe::Types::Contact> >")
    -> {
        "full_type": "std::optional<std::vector<Api::OneMe::Types::Contact>>",
        "type": "Api::OneMe::Types::Contact",
        "optional": True,
        "array": True,
        "map": False,
        "map_key": None,
        "map_value": None,
        "polymorphic": False,
        "polymorphic_base": None,
    }
"""
import template_parser

_QUALIFIERS = ("class ", "struct ", "const ", "volatile ")


def _strip_qualifiers(s):
    """Repeatedly strip leading C++ qualifiers."""
    s = s.strip()
    changed = True
    while changed:
        changed = False
        for pfx in _QUALIFIERS:
            if s.startswith(pfx):
                s = s[len(pfx):].strip()
                changed = True
                break
    return s


def normalize_type(raw):
    """
    Recursively normalize a C++ type string from Binary Ninja's demangled output.

    - Strips class/struct/const/volatile qualifiers at every level.
    - Collapses std::basic_string<char,...> -> std::string
    - Collapses std::basic_string_view<char,...> -> std::string_view
    - Removes allocator/less template args from std::vector/std::map.
    """
    raw = _strip_qualifiers(raw)
    if not raw:
        return raw

    # Fast paths for common simple types
    if "<" not in raw:
        return raw

    # std::basic_string / std::basic_string_view simplification
    if raw.startswith("std::basic_string<"):
        return "std::string"
    if raw.startswith("std::basic_string_view<"):
        return "std::string_view"

    idx = raw.find("<")
    if idx == -1:
        return raw

    tpl_name = raw[:idx].strip()
    rest = raw[idx + 1:]

    # Find matching closing '>' at depth 0
    depth = 0
    close = -1
    for i, c in enumerate(rest):
        if c == "<":
            depth += 1
        elif c == ">":
            if depth == 0:
                close = i
                break
            depth -= 1

    if close == -1:
        return raw  # malformed, return as-is

    inner = rest[:close]
    args = template_parser._extract_tpl_args(inner)
    norm_args = [normalize_type(a) for a in args]

    if tpl_name in ("std::optional", "std::vector",
                     "std::shared_ptr", "std::unique_ptr"):
        if norm_args:
            return "{}<{}>".format(tpl_name, norm_args[0])
        return tpl_name

    if tpl_name in ("std::map", "std::unordered_map"):
        if len(norm_args) >= 2:
            return "{}<{}, {}>".format(tpl_name, norm_args[0], norm_args[1])
        return tpl_name

    if tpl_name == "Api::OneMe::Types::Polymorphic":
        if norm_args:
            return "Api::OneMe::Types::Polymorphic<{}>".format(norm_args[0])
        return tpl_name

    # Generic: keep all args
    return "{}<{}>".format(tpl_name, ", ".join(norm_args))


def _extract_polymorphic_base(s):
    """Return the normalized first arg T from a Polymorphic<T> substring, or None.

    Detection is by substring presence: any occurrence of "Polymorphic<"
    marks the type as polymorphic, regardless of how deeply it is nested
    inside std::optional / std::vector / std::map.
    """
    idx = s.find("Polymorphic<")
    if idx == -1:
        return None
    inner = s[idx + len("Polymorphic<"):]
    first = template_parser._extract_first_tpl_arg(inner)
    return normalize_type(first) if first else None


def decompose_type(raw_type):
    """
    Decompose a raw C++ type string into a structured dict.

    Returns:
        {
            "full_type": str,   # normalized original type
            "type": str | None, # unwrapped base type (None for pure maps)
            "optional": bool,   # wrapped in std::optional<>
            "array": bool,      # wrapped in std::vector<>
            "map": bool,        # is std::map<K,V>
            "map_key": str | None,
            "map_value": str | None,
            "polymorphic": bool,        # contains Api::OneMe::Types::Polymorphic<T>
            "polymorphic_base": str | None,  # inner base type T
        }
    """
    full = normalize_type(raw_type)

    poly_base = _extract_polymorphic_base(full)

    result = {
        "full_type": full,
        "type": full,
        "optional": False,
        "array": False,
        "map": False,
        "map_key": None,
        "map_value": None,
        "polymorphic": poly_base is not None,
        "polymorphic_base": poly_base,
    }

    current = full
    for _ in range(6):  # max nesting depth safety
        current = _strip_qualifiers(current)

        # std::optional<T>
        if current.startswith("std::optional<"):
            inner = _unwrap_single_template(current, "std::optional")
            if inner is not None:
                result["optional"] = True
                current = inner
                continue

        # std::vector<T>
        if current.startswith("std::vector<"):
            inner = _unwrap_single_template(current, "std::vector")
            if inner is not None:
                result["array"] = True
                current = inner
                continue

        # std::shared_ptr<T> / std::unique_ptr<T> — treat as transparent wrapper
        if current.startswith("std::shared_ptr<"):
            inner = _unwrap_single_template(current, "std::shared_ptr")
            if inner is not None:
                current = inner
                continue
        if current.startswith("std::unique_ptr<"):
            inner = _unwrap_single_template(current, "std::unique_ptr")
            if inner is not None:
                current = inner
                continue

        # std::map<K, V, ...> / std::unordered_map<K, V, ...>
        if current.startswith(("std::map<", "std::unordered_map<")):
            k, v = _unwrap_map_args(current)
            if k is not None:
                result["map"] = True
                result["map_key"] = normalize_type(k)
                result["map_value"] = normalize_type(v)
                result["type"] = None
                return result

        break

    result["type"] = normalize_type(current) if current else None
    return result


def _unwrap_single_template(s, prefix):
    """Extract inner type from prefix<T>."""
    s = _strip_qualifiers(s)
    pfx = prefix + "<"
    if not s.startswith(pfx):
        return None
    inner = s[len(pfx):]
    # Strip trailing '>'
    if inner.endswith(">"):
        inner = inner[:-1]
    else:
        # Find matching close
        depth = 0
        for i, c in enumerate(inner):
            if c == "<":
                depth += 1
            elif c == ">":
                if depth == 0:
                    inner = inner[:i]
                    break
                depth -= 1
    return inner.strip()


def _unwrap_map_args(s):
    """Extract (key, value) from std::map<K, V, ...> or std::unordered_map<K, V, ...>."""
    s = _strip_qualifiers(s)
    inner = None
    for pfx in ("std::map<", "std::unordered_map<"):
        if s.startswith(pfx):
            inner = s[len(pfx):]
            break
    if inner is None:
        return None, None
    # Find matching close
    depth = 0
    close = -1
    for i, c in enumerate(inner):
        if c == "<":
            depth += 1
        elif c == ">":
            if depth == 0:
                close = i
                break
            depth -= 1
    if close == -1:
        return None, None
    args = template_parser._extract_tpl_args(inner[:close])
    if len(args) < 2:
        return None, None
    return args[0], args[1]

File: dumper_rust/src/type_parser.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DecomposedType {
    pub full: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub optional: bool,
    pub array: bool,
    pub map: bool,
    pub map_key: Option<String>,
    pub map_value: Option<String>,
    pub polymorphic: bool,
    pub polymorphic_base: Option<String>,
}

const QUALIFIERS: &[&str] = &["class ", "struct ", "enum ", "const ", "volatile "];

pub fn strip_qualifiers(mut s: &str) -> &str {
    s = s.trim();
    let mut changed = true;
    while changed {
        changed = false;
        for pfx in QUALIFIERS {
            if s.starts_with(pfx) {
                s = s[pfx.len()..].trim();
                changed = true;
                break;
            }
        }
    }
    s
}

pub fn extract_template_args(s: &str) -> Vec<String> {
    let mut args = Vec::new();
    let mut depth = 0;
    let mut start = 0;
    for (i, c) in s.char_indices() {
        match c {
            '<' => depth += 1,
            '>' => {
                if depth == 0 {
                    let part = s[start..i].trim();
                    if !part.is_empty() {
                        args.push(part.to_string());
                    }
                    return args;
                }
                depth -= 1;
            }
            ',' => {
                if depth == 0 {
                    let part = s[start..i].trim();
                    if !part.is_empty() {
                        args.push(part.to_string());
                    }
                    start = i + 1;
                }
            }
            _ => {}
        }
    }
    let tail = s[start..].trim();
    if !tail.is_empty() {
        args.push(tail.to_string());
    }
    args
}

pub fn extract_inner_template(s: &str, prefix: &str) -> Option<String> {
    let stripped = strip_qualifiers(s);
    let pfx = format!("{}<", prefix);
    if !stripped.starts_with(&pfx) {
        return None;
    }
    let rest = &stripped[pfx.len()..];
    let mut depth = 0;
    for (i, c) in rest.char_indices() {
        match c {
            '<' => depth += 1,
            '>' => {
                if depth == 0 {
                    return Some(rest[..i].trim().to_string());
                }
                depth -= 1;
            }
            _ => {}
        }
    }
    None
}

pub fn normalize_primitive(s: &str) -> &str {
    match s {
        "int" => "int32_t",
        "__int64" | "long long" => "int64_t",
        "short" => "int16_t",
        "signed char" => "char",
        "unsigned int" => "uint32_t",
        "unsigned __int64" | "unsigned long long" => "uint64_t",
        "unsigned short" => "uint16_t",
        "unsigned char" => "uint8_t",
        other => other,
    }
}

pub fn normalize_type(raw: &str) -> String {
    let s = strip_qualifiers(raw);
    if s.is_empty() {
        return String::new();
    }
    if !s.contains('<') {
        return normalize_primitive(s).to_string();
    }

    if s.starts_with("std::basic_string<") {
        return "std::string".to_string();
    }
    if s.starts_with("std::basic_string_view<") {
        return "std::string_view".to_string();
    }

    let idx = match s.find('<') {
        Some(i) => i,
        None => return s.to_string(),
    };

    let tpl_name = strip_qualifiers(&s[..idx]);
    let rest = &s[idx + 1..];

    let mut depth = 0;
    let mut close = None;
    for (i, c) in rest.char_indices() {
        match c {
            '<' => depth += 1,
            '>' => {
                if depth == 0 {
                    close = Some(i);
                    break;
                }
                depth -= 1;
            }
            _ => {}
        }
    }

    let close_idx = match close {
        Some(i) => i,
        None => return s.to_string(),
    };

    let inner = &rest[..close_idx];
    let args = extract_template_args(inner);
    let norm_args: Vec<String> = args.iter().map(|a| normalize_type(a)).collect();

    match tpl_name {
        "std::optional" | "std::vector" | "std::shared_ptr" | "std::unique_ptr" => {
            if let Some(first) = norm_args.first() {
                format!("{}<{}>", tpl_name, first)
            } else {
                tpl_name.to_string()
            }
        }
        "std::map" | "std::unordered_map" => {
            if norm_args.len() >= 2 {
                format!("{}<{}, {}>", tpl_name, norm_args[0], norm_args[1])
            } else {
                tpl_name.to_string()
            }
        }
        "Api::OneMe::Types::Polymorphic" => {
            if let Some(first) = norm_args.first() {
                format!("Api::OneMe::Types::Polymorphic<{}>", first)
            } else {
                tpl_name.to_string()
            }
        }
        _ => {
            format!("{}<{}>", tpl_name, norm_args.join(", "))
        }
    }
}

pub fn extract_polymorphic_base(s: &str) -> Option<String> {
    let idx = s.find("Polymorphic<")?;
    let inner = &s[idx + "Polymorphic<".len()..];
    let args = extract_template_args(inner);
    let first = args.first()?;
    Some(normalize_type(first))
}

pub fn decompose_type(raw_type: &str) -> DecomposedType {
    let full = normalize_type(raw_type);
    let poly_base = extract_polymorphic_base(&full);

    let mut result = DecomposedType {
        full: full.clone(),
        name: Some(full.clone()),
        optional: false,
        array: false,
        map: false,
        map_key: None,
        map_value: None,
        polymorphic: poly_base.is_some(),
        polymorphic_base: poly_base.clone(),
    };

    let mut current = full;
    for _ in 0..6 {
        let stripped = strip_qualifiers(&current);

        if stripped.starts_with("std::optional<") {
            if let Some(inner) = extract_inner_template(stripped, "std::optional") {
                result.optional = true;
                current = inner;
                continue;
            }
        }

        if stripped.starts_with("std::vector<") {
            if let Some(inner) = extract_inner_template(stripped, "std::vector") {
                result.array = true;
                current = inner;
                continue;
            }
        }

        if stripped.starts_with("std::shared_ptr<") {
            if let Some(inner) = extract_inner_template(stripped, "std::shared_ptr") {
                current = inner;
                continue;
            }
        }

        if stripped.starts_with("std::unique_ptr<") {
            if let Some(inner) = extract_inner_template(stripped, "std::unique_ptr") {
                current = inner;
                continue;
            }
        }

        if stripped.starts_with("Utilities::StrongTypedef<") {
            if let Some(inner) = extract_inner_template(stripped, "Utilities::StrongTypedef") {
                current = inner;
                continue;
            }
        }

        if stripped.starts_with("std::map<") || stripped.starts_with("std::unordered_map<") {
            let pfx = if stripped.starts_with("std::map<") {
                "std::map"
            } else {
                "std::unordered_map"
            };
            if let Some(inner) = extract_inner_template(stripped, pfx) {
                let args = extract_template_args(&inner);
                if args.len() >= 2 {
                    result.map = true;
                    result.map_key = Some(normalize_type(&args[0]));
                    result.map_value = Some(normalize_type(&args[1]));
                    result.name = None;
                    return result;
                }
            }
        }

        break;
    }

    if let Some(base) = poly_base {
        result.name = Some(base);
    } else {
        let norm = normalize_type(&current);
        result.name = if norm.is_empty() { None } else { Some(norm) };
    }

    result
}
